libraries/AP_Scripting/examples/aileron_yaw_mix.lua

```diff
@@ -0,0 +1,197 @@
+end
+
+-- setup quicktune specific parameters
+assert(param:add_table(PARAM_TABLE_KEY, PARAM_TABLE_PREFIX, 5), 'could not add param table')
+
+local AMIX_YAW_FF      = bind_add_param('YAW_FF',        1, 0)
+local AMIX_YAW_P       = bind_add_param('YAW_P',         2, 0)
+local AMIX_YAW_I       = bind_add_param('YAW_I',         3, 0)
+local AMIX_YAW_IMAX    = bind_add_param('YAW_IMAX',      4, 0)
+local AMIX_TELEM_RATE  = bind_add_param('TELEM_RATE',    5, 0)
+
+function constrain(v, minv, maxv)
+   -- constrain a value between two limits
+   if v < minv then
+      return minv
+   end
+   if v > maxv then
+      return maxv
+   end
+   return v
+end
+
+--[[
+ a PI controller with feed-forward implemented as a Lua object, using
+ closure style object
+--]]
+local function PIFF(kFF,kP,kI,iMax)
+   -- the new instance. You can put public variables inside this self
+   -- declaration if you want to
+   local self = {}
+
+   -- private fields as locals
+   local _kFF = kFF
+   local _kP  = kP
+   local _kI  = kI
+
+   local _iMax = iMax
+   local _last_t = nil
+   local _log_data = {}
+   local _I = 0.0
+   local _counter = 0
+
+   local _pid_out = {}
+   _pid_out.FF = 0.0
+   _pid_out.P = 0.0
+   _pid_out.I = 0.0
+   _pid_out.target = 0.0
+   _pid_out.actual = 0.0
+
+   -- update the controller.
+   function self.update(target, current)
+      local now = millis():tofloat() * 0.001
+      if not _last_t then
+         _last_t = now
+      end
+
+      local dt = now - _last_t
+      _last_t = now
+      local err = target - current
+      _counter = _counter + 1
+
+      local FF = _kFF:get() * target
+      local P = _kP:get() * err
+      _I = _I + _kI:get() * err * dt
+      imax = _iMax:get()
+      _I = constrain(_I, -imax, imax)
+      local I = _I
+      local ret = FF + P + I
+
+      _log_data = { target, current, FF, P, I, ret }
+
+      _pid_out.FF = FF
+      _pid_out.P = P
+      _pid_out.I = I
+      _pid_out.target = target
+      _pid_out.actual = current
+
+      return ret
+   end
+
+   -- log the controller internals
+   function self.log(name)
+      logger.write(name,'Targ,Curr,FF,P,I,Total','ffffff',table.unpack(_log_data))
+   end
+
+   function self.send_telemetry()
+      gcs:send_named_float(string.format('YAW_FF'), _pid_out.FF)
+      gcs:send_named_float(string.format('YAW_P'), _pid_out.P)
+      gcs:send_named_float(string.format('YAW_I'), _pid_out.I)
+      gcs:send_named_float(string.format('YAW_TG'), math.deg(_pid_out.target))
+      gcs:send_named_float(string.format('YAW_AC'), math.deg(_pid_out.actual))
+   end
+
+   -- return the instance
+   return self
+end
+
+local yaw_pid = PIFF(AMIX_YAW_FF, AMIX_YAW_P, AMIX_YAW_I, AMIX_YAW_IMAX)
+local last_telem_send = 0.0
+
+function update()
+   -- get the current tilt
+   local tilt_scaled = SRV_Channels:get_output_scaled(k_motor_tilt)
+   if not tilt_scaled then
+      return
+   end
+
+   -- get the current body frame yaw rate
+   local ygyro_rads = ahrs:get_gyro():z()
```

Comment:
Strictly tis is the wrong gyro rate, we need the ahrs view rate. Of course with no Q_TRIM_PITCH or not a tailsitter they should be the same.